webhooks/core/pod_webhook.go

```diff
@@ -22,145 +22,207 @@ import (
+
+	switch WhichPod(pod) {
+	case HostNetworking:
+		response = admission.Allowed("SGP not supported on Pod running on HostNetwork")
+	case Fargate:
+		response = i.HandleFargatePod(req, pod, log)
+	case Linux:
+		response = i.HandleLinuxPod(req, pod, log)
+	case Windows:
+		response = i.HandleWindowsPod(req, pod, log)
+	default:
+		response = admission.Allowed("No criteria met injecting resource limit to Pod")
+	}
+
+	return response
+}
+
+// WhichPod returns the PodType, each PodType will be handled differently
+func WhichPod(pod *corev1.Pod) PodType {
 	// Ignore pod that is scheduled on host network.
 	if pod.Spec.HostNetwork {
-		return admission.Allowed("Pod on HostNetwork will not be injected with resources.")
+		return HostNetworking
+	}
+
+	// This should be the condition before Linux Pods
+	if _, ok := pod.ObjectMeta.Labels[FargatePodIdentifierLabelKey]; ok {
+		return Fargate
+	}
+	// Windows Pod
+	if hasWindowsNodeSelector(pod) {
+		return Windows
+	}
+	return Linux
+}
+
+// HandleFargatePod mutates the Fargate Pod if the Pod Matches a SGP. This also acts like
+// a validation WebHook by removing any existing Annotation on the Pod on Create Event.
+func (i *PodMutationWebHook) HandleFargatePod(req admission.Request, pod *corev1.Pod,
+	log logr.Logger) (response admission.Response) {
+	sgList, err := i.SGPAPI.GetMatchingSecurityGroupForPods(pod)
+	if err != nil {
+		i.Log.Error(err, "failed to get matching SGP for Pods",
+			"namespace", pod.Namespace, "name", pod.Name)
+		return admission.Denied("Failed to get Matching SGP for Pods, rejecting event")
+	}
+
+	switch len(sgList) {
+	case 0:
+		// If Pod is created, with the annotation then this event should be rejected. Only
+		// the controller is allowed to modify this key. This webhook blocks such Create
+		// Events and the validation Webhook blocks all Update events on this key.
+		if _, ok := pod.Annotations[FargatePodSGAnnotationKey]; ok {
+			delete(pod.Annotations, FargatePodSGAnnotationKey)
+			log.Info("Overriding pod-sg annotation added outside of mutating webhook",
+				" Annotation", pod.Annotations)
+			response = i.GetPatchResponse(req, pod, log)
+		} else {
+			// If there's no matching SG for the given Pod
+			response = admission.Allowed("Fargate pod not matching any SGP")
+		}
+	default:
+		// If more than 1 SG match for the Pod then add all matching SG to the Annotation
+		pod.Annotations[FargatePodSGAnnotationKey] = strings.Join(sgList, ",")
+		log.Info("annotating Fargate pod with matching security groups",
+			"Annotations", pod.Annotations)
+		response = i.GetPatchResponse(req, pod, log)
+	}
+
+	return response
+}
+
+// HandleWindowsPod mutates the Windows Pod by injecting a secondary IPv4 Address
+// Limit to the Pod when the Windows IPAM feature is enabled via ConfigMap
+func (i *PodMutationWebHook) HandleWindowsPod(req admission.Request, pod *corev1.Pod,
+	log logr.Logger) (response admission.Response) {
+
+	// TODO: Don't Process event if feature is disabled
```

Comment:
Will this function be called if the feature is disabled at the moment? Is it safe to keep the processing open?

Reply:
We have kept it allow all for now because of integration test. @sushrk is working on a follow up PR that will add the relevant checks here.